core/lib/dal/src/transactions_web3_dal.rs

```diff
@@ -24,171 +22,102 @@ pub struct TransactionsWeb3Dal<'a, 'c> {
-                            NULL::bytea AS "block_hash",
-                            NULL::BIGINT AS "l1_batch_number?",
-                            miniblock_number,
-                            tx_hash,
-                            tx_index_in_block,
-                            event_index_in_block,
-                            event_index_in_tx
-                        FROM
-                            events
-                        WHERE
-                            tx_hash = $1
-                        ORDER BY
-                            miniblock_number ASC,
-                            event_index_in_block ASC
-                        "#,
-                        hash.as_bytes()
-                    )
-                    .instrument("get_transaction_receipt_events")
-                    .with_arg("hash", &hash)
-                    .fetch_all(self.storage.conn())
-                    .await?
+        hashes: &[H256],
+    ) -> Result<Vec<TransactionReceipt>, SqlxError> {
+        let mut receipts: Vec<TransactionReceipt> = sqlx::query_as!(
+            StorageTransactionReceipt,
+            r#"
+            WITH
+                sl AS (
+                    SELECT DISTINCT
+                        ON (storage_logs.tx_hash) *
+                    FROM
+                        storage_logs
+                    WHERE
+                        storage_logs.address = $1
+                        AND storage_logs.tx_hash = ANY ($3)
+                    ORDER BY
+                        storage_logs.tx_hash,
+                        storage_logs.miniblock_number DESC,
+                        storage_logs.operation_number DESC
+                )
+            SELECT
+                transactions.hash AS tx_hash,
+                transactions.index_in_block AS index_in_block,
+                transactions.l1_batch_tx_index AS l1_batch_tx_index,
+                transactions.miniblock_number AS "block_number!",
+                transactions.error AS error,
+                transactions.effective_gas_price AS effective_gas_price,
+                transactions.initiator_address AS initiator_address,
+                transactions.data -> 'to' AS "transfer_to?",
+                transactions.data -> 'contractAddress' AS "execute_contract_address?",
+                transactions.tx_format AS "tx_format?",
+                transactions.refunded_gas AS refunded_gas,
+                transactions.gas_limit AS gas_limit,
+                miniblocks.hash AS "block_hash",
+                miniblocks.l1_batch_number AS "l1_batch_number?",
+                sl.key AS "contract_address?"
+            FROM
+                transactions
+                JOIN miniblocks ON miniblocks.number = transactions.miniblock_number
+                LEFT JOIN sl ON sl.value != $2
+                AND sl.tx_hash = transactions.hash
+            WHERE
+                transactions.hash = ANY ($3)
+            ORDER BY
+                transactions.index_in_block ASC
```

Comment:
Bikeshedding: This ordering looks unnecessarily specialized for the case of loading multiple transactions we have currently. IMO, it'd make more sense to order receipts so that their ordering matches `hashes`, or don't enforce any ordering (clearly documenting this), so that it can be specified by the caller:

```rust
// in `get_block_receipts_impl()`:
let mut receipts = // ...
receipts.sort_unstable_by_key(|receipt| receipt.transaction_index);
// ^ The caller knows that receipts have unique `transaction_index`
```